support cuDNN V5?

Hi @chuckcho 

I compiled this video-caffe with cuDNN v5.0.5 for CUDA 8.0, but got errors like following.

So, does this version of Caffe not support cuDNN v5?

Best,
Rui

```
-- ******************* Caffe Configuration Summary *******************
-- General:
--   Version           :   1.0.0-rc3
--   Git               :   v1.0-rc0-3653-g0089f35-dirty
--   System            :   Linux
--   C++ compiler      :   /usr/bin/c++
--   Release CXX flags :   -O3 -DNDEBUG -fPIC -Wall -Wno-sign-compare -Wno-uninitialized
--   Debug CXX flags   :   -g -fPIC -Wall -Wno-sign-compare -Wno-uninitialized
--   Build type        :   Release
-- 
--   BUILD_SHARED_LIBS :   ON
--   BUILD_python      :   ON
--   BUILD_matlab      :   OFF
--   BUILD_docs        :   ON
--   CPU_ONLY          :   OFF
--   USE_OPENCV        :   ON
--   USE_LEVELDB       :   ON
--   USE_LMDB          :   ON
--   ALLOW_LMDB_NOLOCK :   OFF
-- 
-- Dependencies:
--   BLAS              :   Yes (Atlas)
--   Boost             :   Yes (ver. 1.58)
--   glog              :   Yes
--   gflags            :   Yes
--   protobuf          :   Yes (ver. 2.6.0)
--   lmdb              :   Yes (ver. 0.9.14)
--   LevelDB           :   Yes (ver. 1.18)
--   Snappy            :   Yes (ver. 1.1.3)
--   OpenCV            :   Yes (ver. 3.1.0)
--   CUDA              :   Yes (ver. 8.0)
-- 
-- NVIDIA CUDA:
--   Target GPU(s)     :   Auto
--   GPU arch(s)       :   sm_61
--   cuDNN             :   Yes (ver. 5.0.5)
-- 
-- Python:
--   Interpreter       :   /usr/bin/python2.7 (ver. 2.7.9)
--   Libraries         :   /usr/lib64/libpython2.7.so (ver 2.7.9)
--   NumPy             :   /usr/lib64/python2.7/site-packages/numpy/core/include (ver 1.10.4)
-- 
-- Documentaion:
--   Doxygen           :   /usr/bin/doxygen (1.8.6)
--   config_file       :   /home_local/C3D/new/video-caffe/.Doxyfile
-- 
-- Install:
--   Install path      :   /home_local/C3D/new/video-caffe/build/install
-- 
-- Configuring done
-- Generating done
-- Build files have been written to: /home_local/C3D/new/video-caffe/build
```

`host:/home_local/C3D/new/video-caffe/build> make clean`
`hosr:/home_local/C3D/new/video-caffe/build> make all`

```
[  1%] Running C++/Python protocol buffer compiler on /home_local/C3D/new/video-caffe/src/caffe/proto/caffe.proto
Scanning dependencies of target proto
[  1%] Building CXX object src/caffe/CMakeFiles/proto.dir/__/__/include/caffe/proto/caffe.pb.cc.o
[  1%] Linking CXX static library ../../lib/libproto.a
[  1%] Built target proto
[  1%] Building NVCC (Device) object src/caffe/CMakeFiles/cuda_compile.dir/layers/cuda_compile_generated_eltwise_layer.cu.o
/home_local/C3D/new/video-caffe/include/caffe/util/cudnn.hpp(152): error: argument of type "int *" is incompatible with parameter of type "cudnnTensorFormat_t *"

/home_local/C3D/new/video-caffe/include/caffe/util/cudnn.hpp(153): error: too few arguments in function call

/home_local/C3D/new/video-caffe/include/caffe/util/cudnn.hpp(186): error: argument of type "int" is incompatible with parameter of type "cudnnNanPropagation_t"

/home_local/C3D/new/video-caffe/include/caffe/util/cudnn.hpp(186): error: too few arguments in function call

/home_local/C3D/new/video-caffe/include/caffe/util/cudnn.hpp(209): error: argument of type "std::size_t" is incompatible with parameter of type "cudnnNanPropagation_t"

/home_local/C3D/new/video-caffe/include/caffe/util/cudnn.hpp(209): error: argument of type "int *" is incompatible with parameter of type "int"

/home_local/C3D/new/video-caffe/include/caffe/util/cudnn.hpp(209): error: too few arguments in function call

7 errors detected in the compilation of "/tmp/tmpxft_00007f09_00000000-5_eltwise_layer.cpp4.ii".
CMake Error at cuda_compile_generated_eltwise_layer.cu.o.cmake:266 (message):
  Error generating file
  /home_local/C3D/new/video-caffe/build/src/caffe/CMakeFiles/cuda_compile.dir/layers/./cuda_compile_generated_eltwise_layer.cu.o


src/caffe/CMakeFiles/caffe.dir/build.make:469: recipe for target 'src/caffe/CMakeFiles/cuda_compile.dir/layers/cuda_compile_generated_eltwise_layer.cu.o' failed
make[2]: *** [src/caffe/CMakeFiles/cuda_compile.dir/layers/cuda_compile_generated_eltwise_layer.cu.o] Error 1
CMakeFiles/Makefile2:235: recipe for target 'src/caffe/CMakeFiles/caffe.dir/all' failed
make[1]: *** [src/caffe/CMakeFiles/caffe.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

Some compilation errors with cudnn v5 are as follows.
```
./include/caffe/util/cudnn.hpp: In function ‘void caffe::cudnn::setNdConvolutionDesc(cudnnConvolutionStruct**, cudnnTensorDescriptor_t, cudnnFilterDescriptor_t, std::vector<int>, std::vector<int>)’:
./include/caffe/util/cudnn.hpp:158:19: error: cannot convert ‘int*’ to ‘cudnnTensorFormat_t*’ for argument ‘4’ to ‘cudnnStatus_t cudnnGetFilterNdDescriptor(cudnnFilterDescriptor_t, int, cudnnDataType_t*, cudnnTensorFormat_t*, int*, int*)’
       shape.data());
                   ^
./include/caffe/util/cudnn.hpp: In function ‘void caffe::cudnn::createNdPoolingDesc(cunnPoolingStruct**, caffe::PoolingParameter_PoolMethod, cudnnPoolingMode_t*, std::vector<int>, std::vector<int>, std::vector<int>)’:
./include/caffe/util/cudnn.hpp:220:54: error: too few arguments to function ‘cudnnStatus_t cudnnSetPoolingNdDescriptor(cudnnPoolingDescriptor_t, cudnnPoolingMode_t, cudnnNanPropagation_t, int, const int*, const int*, const int*)’
               shape.data(), pad.data(), stride.data()));
                                                      ^
```

--- a/include/caffe/layers/cudnn_ndconv_layer.hpp
+++ b/include/caffe/layers/cudnn_ndconv_layer.hpp
@@ -61,7 +61,11 @@ class CudnnNdConvolutionLayer : public Layer<Dtype> {
   vector<cudnnConvolutionDescriptor_t> conv_descs_;
   int bottom_offset_, top_offset_, weight_offset_, bias_offset_;
   size_t workspaceSizeInBytes;
-  void *workspace;
+  void** workspace;
+  cudnnConvolutionBwdFilterAlgo_t* bwd_filter_algo_;
+  cudnnConvolutionBwdDataAlgo_t* bwd_data_algo_;
+  size_t* workspace_bwd_filter_sizes_;
+  size_t* workspace_bwd_data_sizes_;
 };
 
 }  // namespace caffe

--- a/include/caffe/util/cudnn.hpp
+++ b/include/caffe/util/cudnn.hpp
@@ -130,8 +130,13 @@ template <typename Dtype>
 inline void createNdFilterDesc(cudnnFilterDescriptor_t* desc,
     std::vector<int> shape) {
   CUDNN_CHECK(cudnnCreateFilterDescriptor(desc));
+#if CUDNN_VERSION_MIN(5, 0, 0)
+  CUDNN_CHECK(cudnnSetFilterNdDescriptor(*desc, dataType<Dtype>::type,
+              CUDNN_TENSOR_NCHW, shape.size(), shape.data()));
+#else
   CUDNN_CHECK(cudnnSetFilterNdDescriptor(*desc, dataType<Dtype>::type,
               shape.size(), shape.data()));
+#endif
 }
 
 template <typename Dtype>
@@ -154,8 +159,14 @@ inline void setNdConvolutionDesc(cudnnConvolutionDescriptor_t* conv,
   int nbDims;
   std::vector<int> shape(pad.size()+2);
   cudnnDataType_t cudnn_type;
+#if CUDNN_VERSION_MIN(5, 0, 0)
+  cudnnTensorFormat_t format;
+  cudnnGetFilterNdDescriptor(filter, shape.size(), &cudnn_type, &format,
+      &nbDims, shape.data());
+#else
   cudnnGetFilterNdDescriptor(filter, shape.size(), &cudnn_type, &nbDims,
       shape.data());
+#endif
   CHECK_EQ(nbDims, pad.size()+2) <<
       "Dimensions of filters and pad don't match !";
   CHECK_EQ(nbDims, stride.size()+2) <<
@@ -216,8 +227,14 @@ inline void createNdPoolingDesc(cudnnPoolingDescriptor_t* pool_desc,
     LOG(FATAL) << "Unknown pooling method.";
   }
   CUDNN_CHECK(cudnnCreatePoolingDescriptor(pool_desc));
+#if CUDNN_VERSION_MIN(5, 0, 0)
+  CUDNN_CHECK(cudnnSetPoolingNdDescriptor(*pool_desc, *mode,
+              CUDNN_NOT_PROPAGATE_NAN, shape.size(), shape.data(), pad.data(),
+              stride.data()));
+#else
   CUDNN_CHECK(cudnnSetPoolingNdDescriptor(*pool_desc, *mode, shape.size(),
               shape.data(), pad.data(), stride.data()));
+#endif
 }
 
 template <typename Dtype>

--- a/scripts/travis/install-deps.sh
+++ b/scripts/travis/install-deps.sh
@@ -104,10 +104,9 @@ if $WITH_CUDA ; then
   ln -s /usr/local/cuda-$CUDA_VERSION /usr/local/cuda
 
   if $WITH_CUDNN ; then
-    # TODO(chuck): for now, stay with cudnn4 until all backward compatibility
-    #              issues are cleared
-    # apt-get install -y --no-install-recommends libcudnn5-dev
-    apt-get install -y --no-install-recommends libcudnn4-dev
+    apt-get install -y --no-install-recommends libcudnn5-dev
+    # TODO: Should both versions be tested to ensure backwards compatibility?
+#    apt-get install -y --no-install-recommends libcudnn4-dev
   fi
 fi
 

--- a/src/caffe/layers/cudnn_ndconv_layer.cpp
+++ b/src/caffe/layers/cudnn_ndconv_layer.cpp
@@ -120,6 +120,11 @@ void CudnnNdConvolutionLayer<Dtype>::LayerSetUp(
   // Create filter descriptor.
   cudnn::createNdFilterDesc<Dtype>(&filter_desc_, weight_shape);
 
+  bwd_filter_algo_= new cudnnConvolutionBwdFilterAlgo_t[bottom.size()];
+  bwd_data_algo_  = new cudnnConvolutionBwdDataAlgo_t[bottom.size()];
+  workspace_bwd_filter_sizes_ = new size_t[bottom.size()];
+  workspace_bwd_data_sizes_ = new size_t[bottom.size()];
+  workspace = new void*[this->group_ * CUDNN_STREAMS_PER_GROUP];
   // Create tensor descriptor(s) for data and corresponding convolution(s).
   for (int i = 0; i < bottom.size(); i++) {
     cudnnTensorDescriptor_t bottom_desc;
@@ -131,6 +136,8 @@ void CudnnNdConvolutionLayer<Dtype>::LayerSetUp(
     cudnnConvolutionDescriptor_t conv_desc;
     cudnn::createConvolutionDesc<Dtype>(&conv_desc);
     conv_descs_.push_back(conv_desc);
+    workspace_bwd_data_sizes_[i] = 0;
+    workspace_bwd_filter_sizes_[i] = 0;
   }
 
   // Tensor descriptor for bias.

--- a/src/caffe/layers/cudnn_ndconv_layer.cu
+++ b/src/caffe/layers/cudnn_ndconv_layer.cu
@@ -80,11 +80,19 @@ void CudnnNdConvolutionLayer<Dtype>::Forward_gpu(
       // Bias.
       if (this->bias_term_) {
         const Dtype* bias_data = this->blobs_[1]->gpu_data();
+#if CUDNN_VERSION_MIN(5, 0, 0)
+        CUDNN_CHECK(cudnnAddTensor(handle_[g],
+                    cudnn::dataType<Dtype>::one,
+                    bias_desc_, bias_data + bias_offset_ * g,
+                    cudnn::dataType<Dtype>::one,
+                    top_descs_[i], top_data + top_offset_ * g));
+#else
         CUDNN_CHECK(cudnnAddTensor_v3(handle_[g],
                     cudnn::dataType<Dtype>::one,
                     bias_desc_, bias_data + bias_offset_ * g,
                     cudnn::dataType<Dtype>::one,
                     top_descs_[i], top_data + top_offset_ * g));
+#endif
       }
     }
 
@@ -126,7 +134,18 @@ void CudnnNdConvolutionLayer<Dtype>::Backward_gpu(
       // Gradient w.r.t. weights.
       if (this->param_propagate_down_[0]) {
         const Dtype* bottom_data = bottom[i]->gpu_data();
-#if CUDNN_VERSION >= 4000
+#if CUDNN_VERSION_MIN(5, 0, 0)
+        CUDNN_CHECK(cudnnConvolutionBackwardFilter(handle_[1*this->group_ +
+                    g],
+                    cudnn::dataType<Dtype>::one,
+                    bottom_descs_[i], bottom_data + bottom_offset_ * g,
+                    top_descs_[i],    top_diff + top_offset_ * g,
+                    conv_descs_[i],
+                    bwd_filter_algo_[i], workspace[1*this->group_ + g],
+                    workspace_bwd_filter_sizes_[i],
+                    cudnn::dataType<Dtype>::one,
+                    filter_desc_, weight_diff + weight_offset_ * g));
+#elif CUDNN_VERSION_MIN(4, 0, 0)
         CUDNN_CHECK(cudnnConvolutionBackwardFilter_v2(handle_[1*this->group_ +
                     g],
                     cudnn::dataType<Dtype>::one,
@@ -153,7 +172,17 @@ void CudnnNdConvolutionLayer<Dtype>::Backward_gpu(
           weight = this->blobs_[0]->gpu_data();
         }
         Dtype* bottom_diff = bottom[i]->mutable_gpu_diff();
-#if CUDNN_VERSION >= 4000
+#if CUDNN_VERSION_MIN(5, 0, 0)
+        CUDNN_CHECK(cudnnConvolutionBackwardData(handle_[2*this->group_ + g],
+                    cudnn::dataType<Dtype>::one,
+                    filter_desc_, weight + weight_offset_ * g,
+                    top_descs_[i], top_diff + top_offset_ * g,
+                    conv_descs_[i],
+                    bwd_data_algo_[i], workspace[1*this->group_ + g],
+                    workspace_bwd_data_sizes_[i],
+                    cudnn::dataType<Dtype>::zero,
+                    bottom_descs_[i], bottom_diff + bottom_offset_ * g));
+#elif CUDNN_VERSION_MIN(4, 0, 0)
         CUDNN_CHECK(cudnnConvolutionBackwardData_v2(handle_[2*this->group_ + g],
                     cudnn::dataType<Dtype>::one,
                     filter_desc_, weight + weight_offset_ * g,

